### ISSUE TITLE
Bolivia parser down

### ISSUE THREAD

**Issue by q--** [tmrowco/electricitymap-contrib#2882]
The page mentioned in DATA_SOURCES gives error 403 Forbidden, so perhaps the API URLs have moved.

[A (new?) page with generation breakdown is here](https://www.cndc.bo/gene/index.php), it appears to have the data the parser used. Haven't investigated if it's in the same format.

---
[Kibana](https://kibana.electricitymap.org/app/kibana#/discover/10af54f0-0c4a-11e9-85c1-1d63df8c862c?_g=(refreshInterval:('$$hashKey':'object:232',display:'5%20minutes',pause:!f,section:2,value:300000),time:(from:now%2Fw,mode:quick,to:now%2Fw))&_a=(columns:!(message,extra.key,level),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!t,index:'96f67170-0c49-11e9-85c1-1d63df8c862c',key:level,negate:!f,params:(query:ERROR,type:phrase),type:phrase,value:ERROR),query:(match:(level:(query:ERROR,type:phrase)))),('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'96f67170-0c49-11e9-85c1-1d63df8c862c',key:extra.key,negate:!f,params:(query:BO,type:phrase),type:phrase,value:BO),query:(match:(extra.key:(query:BO,type:phrase))))),index:'96f67170-0c49-11e9-85c1-1d63df8c862c',interval:auto,query:(language:lucene,query:''),sort:!('@timestamp',desc)))

```
Traceback (most recent call last):
  File "/home/src/electricitymap/feeder/lib/fetch_data.py", line 178, in launch_parsers
    objs = res.get(timeout) or []
  File "/usr/local/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
  File "/usr/local/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/src/electricitymap/feeder/lib/fetch_data.py", line 176, in <lambda>
    **parser_kwargs))
  File "/home/src/electricitymap/contrib/parsers/BO.py", line 204, in fetch_generation_forecast
    obj = webparser(response)
  File "/home/src/electricitymap/contrib/parsers/BO.py", line 43, in webparser
    obj.hour = hours
  File "/usr/local/lib/python3.6/site-packages/pandas/core/generic.py", line 4405, in __setattr__
    self[name] = value
  File "/usr/local/lib/python3.6/site-packages/pandas/core/frame.py", line 3119, in __setitem__
    self._set_item(key, value)
  File "/usr/local/lib/python3.6/site-packages/pandas/core/frame.py", line 3194, in _set_item
    value = self._sanitize_column(key, value)
  File "/usr/local/lib/python3.6/site-packages/pandas/core/frame.py", line 3391, in _sanitize_column
    value = _sanitize_index(value, self.index, copy=False)
  File "/usr/local/lib/python3.6/site-packages/pandas/core/series.py", line 4001, in _sanitize_index
    raise ValueError('Length of values does not match length of ' 'index')
ValueError: Length of values does not match length of index
```

**Comment by q--** [comment]
Status so far:

New URL appears to be ``, still the same error so the old URL might also still work.

The site now requires an XSRF token that changes on every request.
* Is given with the XMLHttpRequest as `X-Csrf-Token`
* Is contained in the page source at https://www.cndc.bo/gene/index.php,  Ctrl+F `ttoken`
* Is also contained in the response JSON, name is again `ttoken`

**Comment by q--** [comment]
Update 2:

* I've written a method to extract the XSRF token. Have implemented it at one spot in the parser. [Progress on this branch if anyone else wants to see or continue work on this](https://github.com/q--/electricitymap-contrib/tree/Fix_BO_parser)


* Old URL appears to be gone for good, even with XSRF token.
* New URL returns data as JSON instead of HTML, which is neat but requires rewriting part of the parser

**Comment by brandongalbraith** [comment]
I think this can be closed? It appears the changes merged in from #2969 work as expected based on live data for the Bolivia area in EM.

![image](https://user-images.githubusercontent.com/350893/122121421-3b432280-cdf1-11eb-9561-3fcfd506ba81.jpeg)

**Comment by brandongalbraith** [comment]
I believe this issue can be closed. It appears the changes merged in from #2969 work as expected based on [live data for the Bolivia area in EM](https://www.electricitymap.org/zone/BO?wind=false&solar=false).

**Comment by skovhus** [comment]
Thanks!
